### PR TITLE
HPO logger handler

### LINKED ISSUE #3020: Scheduled workflow failed
Oh no, something went wrong in the scheduled workflow **PyTorch version tests with commit cf3fdd1739d4df7e2045d9fe54b404da3ef2501f**.
Please look into it:

https://github.com/pytorch/ignite/actions/runs/5827315738

Feel free to close this if this was just a one-off error.

### LINKED ISSUE #3023: Event handler to "unlock" GCPs Vertex.AI hyperparameter tuning service
Dear PyTorch team,

I have developed a custom event handler that would make ignite code (more specifically: MONAI code) accessible to hyperparamer tuning jobs in GCP Vertex AI. It is an Ignite-tified version of this [Link](https://github.com/GoogleCloudPlatform/cloudml-hypertune/blob/master/hypertune/hypertune.py). As you can see from the code in the link, the metrics are simply saved at a certain time, in a certain place with a certain syntax. Once this is possible the Vertex HPO orchestration kicks in. Input arguments are controlled via Vertex.AI custom training jobs and output model performance can be extracted from that output file written at the end of the training.

Why is this useful? With this handler Ignite code can be subject to "outsourced" hyperparameter screening with just adding this handler and a few lines of Vertex config files. I found the outsourcing of HPO to the cloud platform way easier than coding it myself.

If you want I can contribute my solution to the codebase via a PR. Just let me know.
image003_obf
image002_obf
image001

### PR DESCRIPTION
**Pull request by St3V0Bay**
Fixes #3023

Description:

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/contrib/handlers/hpo_logger.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+from enum import Enum
+import json
+import os
+from threading import RLock
+import time
+from typing import Any
+
+from ignite.engine import Engine, Events
+
+_DEFAULT_METRIC_PATH = '/tmp/hypertune/output.metrics'
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:15]: I wonder whether this code could run on windows. If yes, we may need to provide OS-agnostic path using python tempfile module.

File: ignite/contrib/handlers/hpo_logger.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    Args:
+        evaluator: Evaluator to consume metric results from at the end of its evaluation run
+        metric_tag: Converts the metric value coming from the trainer/evaluator's state into a storable value
+    """
+
+    def __init__(
+        self,
+        evaluator: Engine | None = None,
+        metric_tag: UserString = 'training/hptuning/metric'
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:55]: Can't we use `str` here ?
```suggestion
        metric_tag: str = 'training/hptuning/metric'
```

File: ignite/contrib/handlers/hpo_logger.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            evaluator: Ignite Engine implementing network evaluation
+        """
+        evaluator.add_event_handler(Events.EPOCH_COMPLETED, self.log_metrics)
+
+    def __call__(self, engine: Engine) -> None:
+        """
+        Args:
+            engine: Ignite Engine, it can be a trainer, validator or evaluator.
+        """
+        with self.lock:
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:102]: why do we need a lock here ?

File: ignite/contrib/handlers/hpo_logger.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    def state_dict(self):
+        return {MetricLoggerKeys.LOSS: self.loss, MetricLoggerKeys.METRICS: self.metrics}
+
+    def load_state_dict(self, state_dict):
+        self.loss[:] = state_dict[MetricLoggerKeys.LOSS]
+        self.metrics.clear()
+        self.metrics.update(state_dict[MetricLoggerKeys.METRICS])
+
+hpologger = HPOLogger()
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:137]: I think we can remove this line.

File: ignite/contrib/handlers/hpo_logger.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    return output[0]["loss"]
+
+class MetricLoggerKeys(Enum):
+    METRICS = "Metrics"
+    LOSS = "Loss"
+
+
+class HPOLogger(object):
+    """
+    Makes selected metric accessible for use by GCP Vertex AI hyperparameter tuning jobs. By adding only this
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:29]: Can you please add a link to GCP Vertex AI as 
```
`GCP Vertex AI <link>`_
```

File: ignite/contrib/handlers/hpo_logger.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+        # construct the trainer with the logger passed in as a handler so that it logs loss values
+        trainer.add_event_handler(Events.EPOCH_COMPLETED, hpo_logger)
+
+        # run training
+        trainer.run(train_loader, max_epochs=epochs)
+
+    Args:
+        evaluator: Evaluator to consume metric results from at the end of its evaluation run
+        metric_tag: Converts the metric value coming from the trainer/evaluator's state into a storable value
```
Comment (vfdev-5) [inline ignite/contrib/handlers/hpo_logger.py:49]: Below this line we need to add `.. versionadded:: 0.5.0` tag.
See https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#writing-documentation for details

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @St3V0Bay !
Can you please run this script to reformat the code to our format:
```
bash ./tests/run_code_style.sh install
bash ./tests/run_code_style.sh fmt
```
and let's also run mypy:
```
bash ./tests/run_code_style.sh mypy
```
Source: https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md#formatting-without-pre-commit

File: ignite/contrib/handlers/hpo_logger.py (hunk 1 above)
Comment (St3V0Bay) [reply ignite/contrib/handlers/hpo_logger.py:15]: this is the path where the listener of the GCP HPO orchestration has a look. Whatever is dumped here (in terms of metric values) becomes visible to the Vertex platform. I found this path here https://github.com/GoogleCloudPlatform/cloudml-hypertune/blob/8e3530e1c4926ac64cf28330d1104b838e07a468/hypertune/hypertune.py#L24

Should we keep it?
